experimental/dds/tree2/src/simple-tree/proxies.ts

```diff
@@ -87,40 +96,93 @@ export function getProxyForField<TSchema extends TreeFieldSchema>(
+	if (simpleSchemaSymbol in schema) {
+		return schema[simpleSchemaSymbol] as TreeNodeSchemaClass;
+	}
+	return undefined;
+}
+
 export function getOrCreateNodeProxy<TSchema extends TreeNodeSchema>(
-	editNode: FlexTreeNode,
+	flexNode: FlexTreeNode,
 ): TypedNode<TSchema> {
-	const cachedProxy = tryGetEditNodeTarget(editNode);
+	const cachedProxy = tryGetEditNodeTarget(flexNode);
 	if (cachedProxy !== undefined) {
 		return cachedProxy as TypedNode<TSchema>;
 	}
 
-	const schema = editNode.schema;
+	const schema = flexNode.schema;
+	let output: TypedNode<TSchema>;
+	const classSchema = getClassSchema(schema);
+	if (classSchema !== undefined) {
+		if (typeof classSchema === "function") {
+			const simpleSchema = classSchema as unknown as new (
+				dummy: FlexTreeNode,
+			) => TypedNode<TSchema>;
+			output = new simpleSchema(flexNode);
+		} else {
+			output = (
+				schema as unknown as { create: (data: FlexTreeNode) => TypedNode<TSchema> }
+			).create(flexNode);
+		}
+	} else {
+		// Fallback to createNodeProxy if needed.
+		// TODO: maybe remove this fallback and error once migration to class based schema is done.
+		output = createNodeProxy<TSchema>(flexNode, false);
+	}
+	return output;
+}
+
+/**
+ * @param flexNode - underlying tree node which this proxy should wrap.
+ * @param allowAdditionalProperties - If true, setting of unexpected properties will be forwarded to the target object.
+ * Otherwise setting of unexpected properties will error.
+ * @param customTargetObject - Target object of the proxy.
+ * If not provided an empty collection of the relevant type is used for the target and a separate object created to dispatch methods.
+ * If provided, the customTargetObject will be used as both the dispatch object and the proxy target, and therefor must provide needed functionality depending on the schema kind.
+ */
+export function createNodeProxy<TSchema extends TreeNodeSchema>(
+	flexNode: FlexTreeNode,
+	allowAdditionalProperties: boolean,
+	targetObject?: object,
+): TypedNode<TSchema> {
+	const schema = flexNode.schema;
 	if (schemaIsLeaf(schema)) {
-		return editNode.value as TypedNode<TSchema>;
+		return flexNode.value as TypedNode<TSchema>;
 	}
+	let proxy: TypedNode<TSchema>;
 	if (schemaIsMap(schema)) {
-		return setEditNode(
-			createMapProxy(),
-			editNode as FlexTreeMapNode<MapNodeSchema>,
-		) as TypedNode<TSchema>;
+		proxy = createMapProxy(allowAdditionalProperties, targetObject) as TypedNode<TSchema>;
 	} else if (schemaIsFieldNode(schema)) {
-		return setEditNode(
-			createListProxy(),
-			editNode as FlexTreeFieldNode<FieldNodeSchema>,
-		) as TypedNode<TSchema>;
+		proxy = createListProxy(allowAdditionalProperties, targetObject) as TypedNode<TSchema>;
 	} else if (schemaIsObjectNode(schema)) {
-		return setEditNode(
-			createObjectProxy(schema),
-			editNode as FlexTreeObjectNode,
+		proxy = createObjectProxy(
+			schema,
+			allowAdditionalProperties,
+			targetObject,
```

Comment:
@CraigMacomber - I haven't traced this through yet, but passing 'targetObject' makes me think that you're potentially changing the inheritance chain of the proxies?

Before checking this in, I'd recommend updating the compatibility tests under '\src\test\simple-tree' to use SchemaFactory instead of Builder and make sure they still pass.

I believe that they should, because they use '_.object(..)' directly (as opposed to 'class Foo extends _.object(..)'.

If they don't, let me know and I think I can help.

Reply:
Yes, this changes the inheritance change (well allows changing it, which SchemaFactory does), and that was intended. This was the most straightforward way to make methods (inherited properties) added to the class/schema by the schema author work correctly, and also makes instanceof behave correctly.

I suspect it also has the side effect of making trees with different types not compare equal with assert.deepEqual, which I think is a feature, though it also regresses the ability to use node's deep equals to check only the non-type susbest of trees (which might have been a useful feature, or an easy way to fail to compare types on accident: thats subjective).

I'm not clear which exact tests you are referring to. The only references to "compatibility test" in that folder are comments on eslint suppressions. Is there some specific suite you are referring to? The tests for simple-tree are not organize how we normally organize our tests so I'm having some difficulty navigating them.

Reply:
I have added some deepEquals tests for objects wich cover some of these cases.